### ISSUE TITLE
Creating an "auto" selection mode using AIC/Cross-validation

### ISSUE THREAD

**Issue by denizn** [facebook/prophet#549]
Hello,

I have been trying to choose a model manually currently, however, I would like to do this in a more automated fashion in order to remove hyper care supervision I am currently making for each and every forecast.
I want to try and get scores for all of below combinations and then choose accordingly:
multiplicative seasonality vs. additive seasonality
additive promotions (Exreg) vs multiplicative promotions (Exreg)
additive christmas (Exreg) vs multiplicative christmas (Exreg)
additive easter (Exreg) vs multiplicative easter (Exreg)
linear trend vs. logistic trend

Can you please help me how I could get a "Score" for the fit of the data on the historical data?

AIC,BIC,DIC any of these would be good for me. If you have another suggestion or concerns with above that would be even better.

Deniz

**Comment by bletham** [comment]
There are tools for computing various metrics of predictive power (see `performance_metrics` https://github.com/facebook/prophet/blob/v0.3/notebooks/diagnostics.ipynb), so you could do this with the cross validation by choosing the metric you care about (MAPE perhaps) on the forecast horizon that you care about, and then running the cross_validation and calling performance_metrics (in this case with rolling_window=1 to get a single number to use to compare performance of various settings.)

It would be nice to have some built-in tools for doing this type of selection, although writing the loop for the cross_validation described above should not be too onerous.

**Comment by deniznoah** [comment]
Hi Bletham,

Thank you very much for all the information.

I am testing the library to create 100 models/forecasts at a time with different data. For some of them I would like to use log transformation and for some others I do not.

I am trying to use cross_validation results, but with my understanding is that with log transformation and without log transformation the results would be incomparable.

Can you please kindly tell how I would do this with prophet?

Deniz

**Comment by shaidams64** [comment]
Hi, I was wondering if anyone has worked on adding AIC metric for model selection, and if not, can some one help me out where I should start to implement that? Thanks a lot for your help!

**Comment by jpmallette** [comment]
I believe The following suggestion of @bletham is a good start

> There are tools for computing various metrics of predictive power (see performance_metrics https://github.com/facebook/prophet/blob/v0.3/notebooks/diagnostics.ipynb), so you could do this with the cross validation by choosing the metric you care about (MAPE perhaps) on the forecast horizon that you care about, and then running the cross_validation and calling performance_metrics (in this case with rolling_window=1 to get a single number to use to compare performance of various settings.)

I will create a branch auto_cross_validation and propose and implementation with tests.

**Comment by jpmallette** [comment]
Hey, hope you're great. I have written an implementation of @bletham suggestion in a medium blog post [(link here)](https://medium.com/@jeanphilippemallette/prophet-auto-selection-with-cross-validation-7ba2c0a3beef). 

We could add those two functions in a utils.py file with a test_utils.py 
  create_grid : which allow to create a grid of parameters that allow to perform a loop and 
  execute_cross_validation_and_performance_loop : name is self-explanatory

Since it is a loop, I am wondering if this should be built-in or external to Prophet. 
@shaidams64 your feedback is welcome

**Comment by samosun** [comment]
Is AIC in the plan? Since "The choice of these parameters(yearly and weekly seasonnality) could be automated using a model selection procedure such as AIC"(quote from the paper), and I guess AIC could be more practicable than cross validation when we have too few samples to cutoff?

**Comment by samosun** [comment]
> There are tools for computing various metrics of predictive power (see `performance_metrics` https://github.com/facebook/prophet/blob/v0.3/notebooks/diagnostics.ipynb), so you could do this with the cross validation by choosing the metric you care about (MAPE perhaps) on the forecast horizon that you care about, and then running the cross_validation and calling performance_metrics (in this case with rolling_window=1 to get a single number to use to compare performance of various settings.)
> 
> It would be nice to have some built-in tools for doing this type of selection, although writing the loop for the cross_validation described above should not be too onerous.

Hi Bletham,

I am trying to make the choice of yearly seasonality to be automated using AIC, so I think I should get the log-posterior of each model fitting. In fact, I noticed that the "model.optimizing" in forecaster.py returns both the MAP estimation of parameters and the log-posterior, and I guess that's exactly the likelihood to calculate AIC?

@bletham It would be very nice if you could give me some suggestions.

**Comment by samosun** [comment]
@bletham Thank you very much for the reply and the information. 

Now I know I can compute the log likelihood using y_hat and sigma_hat(m.params['sigma_obs']) and compute AIC. Thank you again for your help. 

But as a new learner, I still wonder if I can choose “yearly seasonality” by simply plug so called "log-posterior" output by Stan and #of parameters into AIC formula for computing AIC up to an additive constant?

To be more specific, based on what I understand about your information, the so called "log-posterior" output by Stan is not exact log-posterior because some constants dropped and it's not normalized, it's more like a log(priors multiply by likelihood)up to an additive constant(eg: log (#of sample) term). 

Since other prior scales are fixed when choosing yearly seasonality, is it safe to say that the so called "log-posterior" output by Stan equals log likelihood up to the same additive constant(eg: log (#of sample) term and prior term). If so, then I guess we can have a surrogate for log likelihood in AIC?

**Comment by bohea** [comment]
@bletham  I'm new to prophet and it's great.
I'm using grid_search to do model optimization, and some tests on example dataset suggests that if Prophet(changepoint_prior_scale = 0.01, seasonality_mode='additive') is better than Prophet(changepoint_prior_scale = 0.05, seasonality_mode='additive'), by better I mean lower mape in cross validation, then Prophet(changepoint_prior_scale = 0.01, seasonality_mode='multiplicative' )  must better than Prophet(changepoint_prior_scale = 0.05, seasonality_mode='multiplicative' )

I wonder if this is for sure or just a coincidence

**Comment by bletham** [comment]
No, that seems reasonable to me since these parameters operate fairly independently of each other. The changepoint prior scale determines the trend flexibility, which often will probably have a similar optimum regardless of whether the seasonality on top of that is multiplicative or additive.

 For parameters that don't have strong interactions, I'd expect random search to be much better than grid search. Fig. 1 of this paper provides a great illustration of the benefits of random search over grid search: http://jmlr.csail.mit.edu/papers/volume13/bergstra12a/bergstra12a.pdf

**Comment by bohea** [comment]
thanks i'll try random search

**Comment by jbanik85** [comment]
Not really a Prophet related question. 

@bletham -- do you have any suggestion for any Python library which can be used for random search? I would also appreciate if you could point me towards any sample use case/example (python) that you can recall from top of your mind. I'm fairly new to ML domain, and still struggling a bit to find the right package and approach to hyper parameter tuning.  

Also, do you have any plan to integrate automated hyper parameter tuning in any future version of the Prophet?

**Comment by jbanik85** [comment]
@jpmallette Thanks for your medium post. I left a comment over there. I'd appreciate if you could find some time to respond. Also, if you want, I can cross post the comment, and have the discussion here.

**Comment by bletham** [comment]
We don't intend to include automated hyperparameter tuning.

Random search is really easy, you just generate random values and try them out :-)
Here is a sweep on 10^-3 to 1 (log scale) with 10 iterations
```
cp_prior_vals = np.random.uniform(-3, 0, 10)
seas_prior_vals = np.random.uniform(-3, 0, 10)
for i in range(10):
    m = Prophet(
        seasonality_prior_scale=seas_prior_vals[i],
        changepoint_prior_scale=cp_prior_vals[i],
    ).fit(df)
    # Do cross validation to compute error

**Comment by jbanik85** [comment]
Thanks @bletham . This is very helpful :)

**Comment by bletham** [comment]
What you'd be getting by doing that is the average error across the entire horizon. I think that's probably a good thing to optimize for.

**Comment by vavlopes** [comment]
Just for contribution, i am using random search in my process right now. The number of combinations are quite small but this is a problem i am conscious and it needs to be this way because of the number of time series i have to test.

Packages required:
`
from sklearn.model_selection import ParameterGrid
import itertools
from random import sample
`

 `def create_CV(train_data):
    train_data = train_data.sort_values("ds").reset_index(drop = True)
  
    param_grid = {'growth': ['linear','logistic'],
                'changepoint_prior_scale': list(itertools.chain(*[sample([0.01, 0.1, 0.05],1)],[0.05,0.5,2,10])),
                'changepoint_range': list(np.arange(0.65,0.9,0.05).tolist()),
                'seasonality_prior_scale': list((10**3,100,10,1,0.1,0.01)),
                'holidays_prior_scale' : list((1000,100,10,1,0.1)),
                'capacity':[(train_data.y.max())+30,(train_data.y.max())]
                  }
    #default hyperparameter combination
    default_comb = {'growth': 'linear',
                'changepoint_prior_scale': 0.05,
                'changepoint_range': 0.8,
                'seasonality_prior_scale': 10,
                'holidays_prior_scale' : 10,
                'capacity':(train_data.y.max() + 5)
                  }

    CVGrid = list(ParameterGrid(param_grid))
    i = sample(list(range(len(CVGrid))),40)
    #Using the iths indexes selected to apply a random search
    CVGrid = [CVGrid[id] for id in i]
    #including default combination
    CVGrid.append(default_comb)
    return(CVGrid)`

**Comment by TheXu** [comment]
A bayesian hyper parameter optimization for R Prophet Model is available here:

http://rpubs.com/tdneumann/351073

This is by Terrence Neumann

**Comment by ctselas** [comment]
@bletham, @samosun HI! After checking your comments I tried to use AIC for model evaluation to find the best hyperparameters avoiding CV! I compute the log likelihood as bletham suggests. About the number of parameters of the model, I do something like this:
k = 0
        for param in model.params:
            k += len(param)

But I always have the same output (20). I am not so much familiar with the algorithm so I am not sure if this is the correct way. But I am asking because if it is always the same number, this means that AIC can not be used to find the best model that also generalizes in new data! Please let me know if I am loosing something. Thank you in advance!

**Comment by leophin** [comment]
> Just for contribution, i am using random search in my process right now. The number of combinations are quite small but this is a problem i am conscious and it needs to be this way because of the number of time series i have to test.
> 
> The below function returns a list of dicts. Each of those dicts can be passed as **kwargs on prohet. Hope it helps.
> 
> Packages required:
> `from sklearn.model_selection import ParameterGrid ; import itertools ; from random import sample `
> 
> ```
> def create_CV(train_data):
>     train_data = train_data.sort_values("ds").reset_index(drop = True)
>   
>     param_grid = {'growth': ['linear','logistic'],
>                 'changepoint_prior_scale': list(itertools.chain(*[sample([0.01, 0.1, 0.05],1)],[0.05,0.5,2,10])),
>                 'changepoint_range': list(np.arange(0.65,0.9,0.05).tolist()),
>                 'seasonality_prior_scale': list((10**3,100,10,1,0.1,0.01)),
>                 'holidays_prior_scale' : list((1000,100,10,1,0.1)),
>                 'capacity':[(train_data.y.max())+30,(train_data.y.max())]
>                   }
>     #Default combination of hyperparameters
>     default_comb = {'growth': 'linear',
>                 'changepoint_prior_scale': 0.05,
>                 'changepoint_range': 0.8,
>                 'seasonality_prior_scale': 10,
>                 'holidays_prior_scale' : 10,
>                 'capacity':(train_data.y.max() + 5)
>                   }
> 
>     CVGrid = list(ParameterGrid(param_grid))
>     i = sample(list(range(len(CVGrid))),40)
>     #Using the iths indexes selected to apply a random search
>     CVGrid = [CVGrid[id] for id in i]
>     #including default combination
>     CVGrid.append(default_comb)
>     return(CVGrid)
> ```

can you please tell me how can i find best parameter
